### PR TITLE
PPO Performance Improvements

### LINKED ISSUE #1425: [RFC] PPO Performance Optimizations (or: PPOPO)
We provide many wonderful examples of using `torch.compile` in our repo. Integrating compile into the RLHF recipe could significantly improve performance. It'd also be a unique selling point of the recipe.

In general, should we try and compile the largest chunks of code as possible so as to provide as much opportunity as possible for the compiler to optimize? If so, I'll perhaps re-order the suggestions below as 1) compile generation 2) compile whole-trajectory generation 3) compile loss step.

Wr.t. trajectory generation: trajectory generation involves:
```python
"""
        1: Generate responses, and logits corresponding to the responses using the current policy,
            generating (query, response) pairs.
        2. Estimate logprobs of the generated responses using the current policy.
        3. Estimate values from the generated responses using the current value function.
        4. Replace any tokens in the response after the first stop token (usually EOS token) with padding,
            producting truncated responses.
        5. Run the reward model on the (query, truncated-response) pairs.
        6. Mask out all the invalid values in the trajectory due to padding tokens.
"""
```
There are two options here:
1) Refactor step 1), the generation step, and use kv-cache generation with `generate_next_token` compiled. Then, we compile steps 2-6 in one function, separately.
2) Compile all steps 1-6 without compiling `generate_next_token` separately. 

Note above: I think in gpt-fast both the prefill step, and `generate_next_token` are compiled, separately.

### The plan (subject to above)

1) Let's completely disable compile in the recipe. Currently, compiling our policy and value models causes significant recompiles since tensors are switching between inference and training modes. Obtain a sensible benchmark here vs default compile settings.
2) Integrate #1424 into the recipe, and enable KV cache generation - compare against above (using whichever global compile config is the fastest).
3) As above, but with compiled generation. 2 and 3 alone should provide order(s?) of magnitude speedup in trajectory generation.
6) Compile loss step
7) Compile the entire trajectory generation step?

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
Closes #1425 

This PR provides various performance improvements to our PPO single device recipe. 

| Branch          | Total training time (hours)* | Peak memory allocated (GB) |
| --------------- | ---------------------------- | -------------------------- |
| Main            | 13.1                         | 69.6                       |
| This branch            |       5.4                       |               69.5             |
| This branch + compile | 4.6                          | 68.6                       |

*The models were trained over approx. 37M tokens (~65k samples w/`max_seq_len=512`) on a single A100 GPU. 

image
Due to the non-determinism of the training process curves may look slightly different.

Changelog:
- KV-cacheing is now supported during trajectory generation - this significantly speeds up training.
- `generation.generate` now only returns logits over the *generated* tokens rather than the whole sequence - significantly reduces peak memory usage. Tests have been updated.
- Added profiler support to the recipe.
- Various changes in trajectory estimation/reward estimation which improve performance.
- Added `parents=True` to `output_dir.mkdir` in our checkpointers. We use nested checkpoint folders for PPO e..g `output_dir/policy/`, `output_dir/value/`.
- The addition of various performance improvements in main since the original baseline means we can bump the default batch size in the configs.
- Compile support. We have two options here:
    1) Compile the trajectory estimation functions separately - minimizes recompiles but results in a small warmup overhead.
    2) Compile each model using `training.compile_model` - this results in ~10 recompile warnings, which means we need to increase the compile cache size limit - I've added `torch._dynamo.config.cache_size_limit = 16` at the top of the recipe. 

Pasted image 20241125141622

I landed on option 2 - it's similar to how we integrate compile with the rest of our recipes, and it eliminates the small warmup overhead. To fully realize compile speedups it's recommended to do a small warm-run of the recipe with compile enabled.

### CODE REVIEW

**Comment by felipemello1** [comment]
Geez! >3x improvement is no joke. I don't think i will have time to review it this week. But I am very curious to see the changes.

File: torchtune/modules/transformer.py [hunk 1]
```diff
@@ -418,12 +418,11 @@ def setup_caches(
             encoder_max_seq_len (Optional[int]): maximum encoder cache sequence length.
             decoder_max_seq_len (Optional[int]): maximum decoder cache sequence length.
         """
-
         has_encoder_layers = any(
             isinstance(m, TransformerCrossAttentionLayer) for m in self.modules()
         )
         has_decoder_layers = any(
-            isinstance(l, TransformerSelfAttentionLayer) for l in self.layers
+            isinstance(m, TransformerSelfAttentionLayer) for m in self.modules()
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/modules/transformer.py:425]: I think it will also exist as a property of a `FusionLayer` in vision models

File: torchtune/rlhf/loss/ppo.py [hunk 2]
```diff
@@ -102,7 +104,7 @@ def forward(
             phi_old_values + self.value_clip_range,
         )
         value_loss = torch.maximum(
-            (phi_values - returns) ** 2, (values_clipped - returns) ** 2
+            (phi_values - returns).pow(2), (values_clipped - returns).pow(2)
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/rlhf/loss/ppo.py:107]: I believe so - you don't like? I'm not that bothered if there's stylistic preferences - I prefer `.pow`

File: torchtune/rlhf/rewards.py [hunk 3]
```diff
@@ -113,39 +109,28 @@ def masked_mean(
     Returns:
         torch.Tensor: The mean tensor.
     """
-    return (x * mask).sum(dim=dim) / mask.sum(dim=dim)
+    return (x * mask).sum(dim=dim) / (mask.sum(dim=dim) + 1e-8)
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/rlhf/rewards.py:112]: If (for some reason) all of the elements in the mask are `False` - it's just an annoying edge case

File: torchtune/generation/_generation.py [hunk 4]
```diff
@@ -189,7 +189,7 @@ def get_position_ids_from_padding_mask(
     return ((padding_mask.cumsum(-1) - 1) * padding_mask).to(torch.int)
 
 
-@torch.inference_mode()
+@torch.no_grad()
```
[the remark this answers is not in the document]
Comment (joecummings) [reply torchtune/generation/_generation.py:192]: Is it possible to define "without really being that useful"?

File: torchtune/rlhf/loss/ppo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (joecummings) [reply torchtune/rlhf/loss/ppo.py:107]: Haha yeah, I'm just a cranky old man who likes my double asterisk Python.

File: torchtune/generation/_generation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/generation/_generation.py:192]: I'll need to double check where I found this - I think it was in a PyTorch dev podcast, but when it was released PyTorch folks mentioned up to ~5% improvement gains on deployed models internally at FB. HF PRs for including `inference_mode` in generation didn't really find speedups to the same degree, so they still use `no_grad` for generation. To expand on my point above, under compile `inference_mode` tensors have different metadata properties and we trigger recompiles when guards are created on these properties which results in increased warmup time.

File: torchtune/generation/_generation.py [hunk 5]
```diff
@@ -355,8 +355,8 @@ def generate(
         # if incremental decoding is enabled, we can use the current position
         # otherwise, we take the whole sequence up to the current position
         if incremental_decoding:
-            curr_input_pos = input_pos[:, curr_pos]
-            curr_masks = masks[:, curr_pos, None, :]
+            curr_input_pos = input_pos[:, curr_pos].contiguous()
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/generation/_generation.py:358]: It's still pretty minimal but YMMV depending on bsz and sequence length. When I last profiled it:

image

The little black sliver is the `.contiguous` call, which takes around 50us every step (some napkin math means this overhead is roughly  `50us * max_generated_tokens - 1`) compared to the 4.562s for generating the entire sequence, so a minimal portion of time.

File: torchtune/rlhf/rewards.py [hunk 6]
```diff
@@ -113,39 +109,28 @@ def masked_mean(
...
     if unbiased:
-        mask_sum = mask.sum()
-        if mask_sum == 0:
-            raise ValueError(
-                "The sum of the mask is zero, which can happen when ``ppo_batch_size=1``;"
-                "try increase the ``ppo_batch_size`` or ``gradient_accumulation_steps``"
-            )
-        # note that if mask_sum == 1, then there is a division by zero issue
-        # to avoid it you just need to use a larger minibatch_size
+        mask_sum = mask.sum() + 1e-8
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/rlhf/rewards.py:133]: I'm not sure when someone would want to consider configuring this value

File: torchtune/generation/_generation.py [hunk 7]
```diff
@@ -94,15 +94,15 @@ def generate_next_token(
             - tokens (torch.Tensor): tensor with the generated tokens,
                 with shape [bsz x 1].
             - logits (torch.Tensor): tensor with the logits associated with the generated tokens,
-                with shape [bsz x seq_length x vocab_size].
+                with shape [bsz x 1 x vocab_size].
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/generation/_generation.py:97]: I'm afraid this is going to be challenging to do without introducing graph breaks during compilation. I generally agree with you, though in this case I'm not 100% sure who would be using logins returned from this function outside of PPO

**Comment by SalmanMohammadi** [comment]
Is there anything else blocking this from landing? @joecummings
